File: plataforma-geointeligente-bakc/Services/DashboardService/DashboardService.cs
using Microsoft.Data.SqlClient;
using Microsoft.EntityFrameworkCore;
using SaludPublicaBackend.Configurations.Databases;
using SaludPublicaBackend.Dtos.DashboardDto;
using System.Data;
using System.Data.Common;

namespace SaludPublicaBackend.Services.DashboardService
{
    /// <summary>
    /// Implementación del servicio de dashboard.
    /// Ejecuta directamente las vistas y procedimientos almacenados de la BD
    /// usando ADO.NET sobre la conexión de EF Core.
    /// </summary>
    public class DashboardService : IDashboardService
    {
        private readonly AppDbContext _context;

        public DashboardService(AppDbContext context)
        {
            _context = context;
        }

        #region ── Helpers ADO.NET ─────────────────────────────

        private static double SafeDouble(object val)
            => val == DBNull.Value ? 0.0 : Convert.ToDouble(val);

        private static int SafeInt(object val)
            => val == DBNull.Value ? 0 : Convert.ToInt32(val);

        private static long SafeLong(object val)
            => val == DBNull.Value ? 0L : Convert.ToInt64(val);

        private static string SafeStr(object val)
            => val == DBNull.Value ? string.Empty : val.ToString()!;

        private static DateTime SafeDt(object val)
            => val == DBNull.Value ? DateTime.MinValue : Convert.ToDateTime(val);

        private static decimal? SafeDecimalNull(object val)
            => val == DBNull.Value ? null : Convert.ToDecimal(val);

        private static long? SafeLongNull(object val)
            => val == DBNull.Value ? null : Convert.ToInt64(val);

        private static SqlParameter P(string name, object? value, SqlDbType type, int size = 0)
        {
            var p = size > 0 ? new SqlParameter(name, type, size) : new SqlParameter(name, type);
            p.Value = value ?? DBNull.Value;
            return p;
        }

        /// <summary>Ejecuta un SELECT sobre una vista y mapea todas las filas.</summary>
        private async Task<List<T>> ViewListAsync<T>(string sql, Func<DbDataReader, T> map)
        {
            var list = new List<T>();
            var conn = _context.Database.GetDbConnection();
            var wasOpen = conn.State == ConnectionState.Open;
            if (!wasOpen) await conn.OpenAsync();
            try
            {
                await using var cmd = conn.CreateCommand();
                cmd.CommandText = sql;
                cmd.CommandType = CommandType.Text;
                cmd.CommandTimeout = 60;

                await using var rdr = await cmd.ExecuteReaderAsync();
                while (await rdr.ReadAsync()) list.Add(map(rdr));
            }
            finally { if (!wasOpen) await conn.CloseAsync(); }
            return list;
        }

        /// <summary>Ejecuta un SP y mapea todas las filas.</summary>
        private async Task<List<T>> SpListAsync<T>(
            string sp, Action<List<SqlParameter>> addParams, Func<DbDataReader, T> map)
        {
            var list = new List<T>();
            var conn = _context.Database.GetDbConnection();
            var wasOpen = conn.State == ConnectionState.Open;
            if (!wasOpen) await conn.OpenAsync();
            try
            {
                await using var cmd = conn.CreateCommand();
                cmd.CommandText = sp;
                cmd.CommandType = CommandType.StoredProcedure;
                cmd.CommandTimeout = 120;
                var parms = new List<SqlParameter>();
                addParams(parms);
                foreach (var p in parms) cmd.Parameters.Add(p);

                await using var rdr = await cmd.ExecuteReaderAsync();
                while (await rdr.ReadAsync()) list.Add(map(rdr));
            }
            finally { if (!wasOpen) await conn.CloseAsync(); }
            return list;
        }

        /// <summary>Ejecuta un SP y mapea la primera fila (o default).</summary>
        private async Task<T?> SpSingleAsync<T>(
            string sp, Action<List<SqlParameter>> addParams, Func<DbDataReader, T> map) where T : class
        {
            var conn = _context.Database.GetDbConnection();
            var wasOpen = conn.State == ConnectionState.Open;
            if (!wasOpen) await conn.OpenAsync();
            try
            {
                await using var cmd = conn.CreateCommand();
                cmd.CommandText = sp;
                cmd.CommandType = CommandType.StoredProcedure;
                cmd.CommandTimeout = 120;
                var parms = new List<SqlParameter>();
                addParams(parms);
                foreach (var p in parms) cmd.Parameters.Add(p);

                await using var rdr = await cmd.ExecuteReaderAsync();
                return await rdr.ReadAsync() ? map(rdr) : null;
            }
            finally { if (!wasOpen) await conn.CloseAsync(); }
        }

        /// <summary>Ejecuta un SP que devuelve dos result-sets (tabla paginada).</summary>
        private async Task<TablaDatosResponseDto> SpTablaPaginadaAsync(
            string sp, Action<List<SqlParameter>> addParams)
        {
            var response = new TablaDatosResponseDto();
            var conn = _context.Database.GetDbConnection();
            var wasOpen = conn.State == ConnectionState.Open;
            if (!wasOpen) await conn.OpenAsync();
            try
            {
                await using var cmd = conn.CreateCommand();
                cmd.CommandText = sp;
                cmd.CommandType = CommandType.StoredProcedure;
                cmd.CommandTimeout = 120;
                var parms = new List<SqlParameter>();
                addParams(parms);
                foreach (var p in parms) cmd.Parameters.Add(p);

                await using var rdr = await cmd.ExecuteReaderAsync();

                // Primer result-set: TotalRegistros
                if (await rdr.ReadAsync())
                    response.TotalRegistros = SafeLong(rdr["TotalRegistros"]);

                // Segundo result-set: filas de datos
                if (await rdr.NextResultAsync())
                {
                    while (await rdr.ReadAsync())
                    {
                        response.Datos.Add(new TablaDatosFilaDto
                        {
                            IdRegistro = SafeInt(rdr["idRegistro"]),
                            FechaIngreso = SafeDt(rdr["fechaIngreso"]),
                            Anio = SafeInt(rdr["Anio"]),
                            DiasEstancia = SafeInt(rdr["diasEstancia"]),
                            NombreEstado = SafeStr(rdr["nombreEstado"]),
                            NombreMunicipio = SafeStr(rdr["nombreMunicipio"]),
                            Edad = SafeInt(rdr["edad"]),
                            Sexo = SafeStr(rdr["Sexo"]),
                            CodigoGrupo = SafeStr(rdr["CodigoGrupo"]),
                            CodigoICD = SafeStr(rdr["codigoICD"]),
                            ClaveInstitucion = SafeStr(rdr["ClaveInstitucion"]),
                            NombreInstitucion = SafeStr(rdr["NombreInstitucion"]),
                            EstratoUnidad = SafeStr(rdr["EstratoUnidad"]),
                            IdMotivoEgreso = SafeInt(rdr["idMotivoEgreso"]),
                            EsDefuncion = SafeInt(rdr["EsDefuncion"])
                        });
                    }
                }
            }
            finally { if (!wasOpen) await conn.CloseAsync(); }
            return response;
        }

        /// <summary>Agrega los parámetros comunes de filtro a un SP del dashboard.</summary>
        private static void AddFiltrosComunes(List<SqlParameter> parms, DashboardFiltrosDto f)
        {
            parms.Add(P("@AnioInicio", f.AnioInicio, SqlDbType.Int));
            parms.Add(P("@AnioFin", f.AnioFin, SqlDbType.Int));
            parms.Add(P("@CodigoGrupo", f.CodigoGrupo, SqlDbType.VarChar, 255));
            parms.Add(P("@CodigoICD", f.CodigoICD, SqlDbType.VarChar, 50));
            parms.Add(P("@IdEstado", f.IdEstado, SqlDbType.Int));
            parms.Add(P("@IdSexo", f.IdSexo, SqlDbType.Int));
            parms.Add(P("@IdRangoEdad", f.IdRangoEdad, SqlDbType.Int));
            parms.Add(P("@ClaveInstitucion", f.ClaveInstitucion, SqlDbType.NVarChar, 50));
            parms.Add(P("@Estrato", f.Estrato, SqlDbType.NVarChar, 50));
        }

        #endregion

        // ═══════════════════════════════════════════════════════
        //  CATÁLOGOS / FILTROS  (vistas + SPs de catálogo)
        // ═══════════════════════════════════════════════════════

        public Task<List<AnioDisponibleDto>> GetAniosDisponiblesAsync()
            => ViewListAsync(
                "SELECT Anio FROM dbo.vw_AniosDisponibles ORDER BY Anio",
                r => new AnioDisponibleDto { Anio = SafeInt(r["Anio"]) });

        public Task<List<CategoriaEnfermedadDto>> GetCategoriasBaseAsync()
            => SpListAsync("dbo.sp_ObtenerCategoriasBase",
                _ => { },
                r => new CategoriaEnfermedadDto
                {
                    CodigoGrupo = SafeStr(r["CodigoGrupo"]),
                    CodigoICD = SafeStr(r["codigoICD"]),
                    NombreEnfermedad = SafeStr(r["nombreEnfermedad"])
                });

        public Task<List<SubcategoriaEnfermedadDto>> GetSubcategoriasAsync(string codigoGrupo)
            => SpListAsync("dbo.sp_ObtenerSubcategorias",
                p => p.Add(P("@CodigoGrupo", codigoGrupo, SqlDbType.VarChar, 3)),
                r => new SubcategoriaEnfermedadDto
                {
                    CodigoICD = SafeStr(r["codigoICD"]),
                    NombreEnfermedad = SafeStr(r["nombreEnfermedad"])
                });

        public Task<List<EstadoActivoDto>> GetEstadosActivosAsync()
            => ViewListAsync(
                "SELECT idEstado, nombreEstado FROM dbo.vw_EstadosActivos ORDER BY nombreEstado",
                r => new EstadoActivoDto
                {
                    IdEstado = SafeInt(r["idEstado"]),
                    NombreEstado = SafeStr(r["nombreEstado"])
                });

        public Task<List<RangoEdadFiltroDto>> GetRangosEdadAsync()
            => ViewListAsync(
                "SELECT idRangoEdad, EtiquetaRango FROM dbo.vw_FiltroRangoEdad ORDER BY idRangoEdad",
                r => new RangoEdadFiltroDto
                {
                    IdRangoEdad = SafeInt(r["idRangoEdad"]),
                    EtiquetaRango = SafeStr(r["EtiquetaRango"])
                });

        public Task<List<SexoFiltroDto>> GetSexosAsync()
            => ViewListAsync(
                "SELECT idSexo, descripcion FROM dbo.vw_FiltroSexo ORDER BY idSexo",
                r => new SexoFiltroDto
                {
                    IdSexo = SafeInt(r["idSexo"]),
                    Descripcion = SafeStr(r["descripcion"])
                });

        public Task<List<InstitucionFiltroDto>> GetInstitucionesAsync()
            => ViewListAsync(
                "SELECT ClaveInstitucion, Institucion FROM dbo.vw_Instituciones ORDER BY ClaveInstitucion",
                r => new InstitucionFiltroDto
                {
                    ClaveInstitucion = SafeStr(r["ClaveInstitucion"]),
                    Institucion = SafeStr(r["Institucion"])
                });

        public Task<List<EstratoFiltroDto>> GetEstratosAsync()
            => ViewListAsync(
                "SELECT EstratoUnidad FROM dbo.vw_Estratos ORDER BY EstratoUnidad",
                r => new EstratoFiltroDto
                {
                    EstratoUnidad = SafeStr(r["EstratoUnidad"])
                });

        // ═══════════════════════════════════════════════════════
        //  INDICADORES TOTALES  →  sp_Dashboard_IndicadoresTotales
        // ═══════════════════════════════════════════════════════

        public async Task<IndicadoresTotalesDto> GetIndicadoresTotalesAsync(DashboardFiltrosDto filtros)
        {
            var dto = await SpSingleAsync("dbo.sp_Dashboard_IndicadoresTotales",
                p => AddFiltrosComunes(p, filtros),
                r => new IndicadoresTotalesDto
                {
                    TotalCasos = SafeLong(r["TotalCasos"]),
                    TotalDefunciones = SafeLong(r["TotalDefunciones"]),
                    PoblacionBase = SafeLong(r["PoblacionBase"]),
                    TasaIncidencia = SafeDecimalNull(r["TasaIncidencia"]),
                    TasaMortalidad = SafeDecimalNull(r["TasaMortalidad"])
                });
            return dto ?? new IndicadoresTotalesDto();
        }

        // ═══════════════════════════════════════════════════════
        //  TENDENCIA ANUAL  →  sp_Dashboard_GraficaTendencia
        // ═══════════════════════════════════════════════════════

        public Task<List<TendenciaAnualDto>> GetTendenciaAsync(DashboardFiltrosDto filtros)
            => SpListAsync("dbo.sp_Dashboard_GraficaTendencia",
                p => AddFiltrosComunes(p, filtros),
                r => new TendenciaAnualDto
                {
                    Anio = SafeInt(r["Anio"]),
                    TotalCasos = SafeLong(r["TotalCasos"]),
                    TotalDefunciones = SafeLong(r["TotalDefunciones"]),
                    Poblacion = SafeLongNull(r["Poblacion"]),
                    TasaIncidencia = SafeDecimalNull(r["TasaIncidencia"]),
                    TasaMortalidad = SafeDecimalNull(r["TasaMortalidad"])
                });

        // ═══════════════════════════════════════════════════════
        //  MAPA DE ESTADOS  →  sp_Dashboard_MapaEstados
        //  (no recibe @IdEstado — es mapa de todos los estados)
        // ═══════════════════════════════════════════════════════

        public Task<List<MapaEstadoDto>> GetMapaEstadosAsync(DashboardFiltrosDto filtros)
            => SpListAsync("dbo.sp_Dashboard_MapaEstados",
                p =>
                {
                    // El SP del mapa no recibe @IdEstado (siempre todos los estados)
                    p.Add(P("@AnioInicio", filtros.AnioInicio, SqlDbType.Int));
                    p.Add(P("@AnioFin", filtros.AnioFin, SqlDbType.Int));
                    p.Add(P("@CodigoGrupo", filtros.CodigoGrupo, SqlDbType.VarChar, 255));
                    p.Add(P("@CodigoICD", filtros.CodigoICD, SqlDbType.VarChar, 50));
                    p.Add(P("@IdSexo", filtros.IdSexo, SqlDbType.Int));
                    p.Add(P("@IdRangoEdad", filtros.IdRangoEdad, SqlDbType.Int));
                    p.Add(P("@ClaveInstitucion", filtros.ClaveInstitucion, SqlDbType.NVarChar, 50));
                    p.Add(P("@Estrato", filtros.Estrato, SqlDbType.NVarChar, 50));
                },
                r => new MapaEstadoDto
                {
                    IdEstado = SafeInt(r["idEstado"]),
                    NombreEstado = SafeStr(r["nombreEstado"]),
                    Casos = SafeLong(r["Casos"]),
                    Defunciones = SafeLong(r["Defunciones"]),
                    Poblacion = SafeLongNull(r["Poblacion"]),
                    TasaIncidencia = SafeDecimalNull(r["TasaIncidencia"]),
                    TasaMortalidad = SafeDecimalNull(r["TasaMortalidad"])
                });

        // ═══════════════════════════════════════════════════════
        //  GRÁFICA POR RANGO DE EDAD  →  sp_Dashboard_GraficaEdades
        //  (no recibe @IdRangoEdad — muestra todos los rangos)
        // ═══════════════════════════════════════════════════════

        public Task<List<GraficaEdadDto>> GetGraficaEdadesAsync(DashboardFiltrosDto filtros)
            => SpListAsync("dbo.sp_Dashboard_GraficaEdades",
                p =>
                {
                    p.Add(P("@AnioInicio", filtros.AnioInicio, SqlDbType.Int));
                    p.Add(P("@AnioFin", filtros.AnioFin, SqlDbType.Int));
                    p.Add(P("@CodigoGrupo", filtros.CodigoGrupo, SqlDbType.VarChar, 255));
                    p.Add(P("@CodigoICD", filtros.CodigoICD, SqlDbType.VarChar, 50));
                    p.Add(P("@IdEstado", filtros.IdEstado, SqlDbType.Int));
                    p.Add(P("@IdSexo", filtros.IdSexo, SqlDbType.Int));
                    p.Add(P("@ClaveInstitucion", filtros.ClaveInstitucion, SqlDbType.NVarChar, 50));
                    p.Add(P("@Estrato", filtros.Estrato, SqlDbType.NVarChar, 50));
                },
                r => new GraficaEdadDto
                {
                    IdRangoEdad = SafeInt(r["idRangoEdad"]),
                    RangoEdad = SafeStr(r["RangoEdad"]),
                    TotalCasos = SafeLong(r["TotalCasos"]),
                    TotalDefunciones = SafeLong(r["TotalDefunciones"])
                });

        // ═══════════════════════════════════════════════════════
        //  GRÁFICA POR INSTITUCIÓN  →  sp_Dashboard_GraficaInstituciones
        //  (no recibe @ClaveInstitucion)
        // ═══════════════════════════════════════════════════════

        public Task<List<GraficaInstitucionDto>> GetGraficaInstitucionesAsync(DashboardFiltrosDto filtros)
            => SpListAsync("dbo.sp_Dashboard_GraficaInstituciones",
                p =>
                {
                    p.Add(P("@AnioInicio", filtros.AnioInicio, SqlDbType.Int));
                    p.Add(P("@AnioFin", filtros.AnioFin, SqlDbType.Int));
                    p.Add(P("@CodigoGrupo", filtros.CodigoGrupo, SqlDbType.VarChar, 255));
                    p.Add(P("@CodigoICD", filtros.CodigoICD, SqlDbType.VarChar, 50));
                    p.Add(P("@IdEstado", filtros.IdEstado, SqlDbType.Int));
                    p.Add(P("@IdSexo", filtros.IdSexo, SqlDbType.Int));
                    p.Add(P("@IdRangoEdad", filtros.IdRangoEdad, SqlDbType.Int));
                    p.Add(P("@Estrato", filtros.Estrato, SqlDbType.NVarChar, 50));
                },
                r => new GraficaInstitucionDto
                {
                    ClaveInstitucion = SafeStr(r["ClaveInstitucion"]),
                    NombreInstitucion = SafeStr(r["NombreInstitucion"]),
                    TotalCasos = SafeLong(r["TotalCasos"]),
                    TotalDefunciones = SafeLong(r["TotalDefunciones"])
                });

        // ═══════════════════════════════════════════════════════
        //  GRÁFICA POR ESTRATO  →  sp_Dashboard_GraficaEstratos
        //  (no recibe @Estrato)
        // ═══════════════════════════════════════════════════════

        public Task<List<GraficaEstratoDto>> GetGraficaEstratosAsync(DashboardFiltrosDto filtros)
            => SpListAsync("dbo.sp_Dashboard_GraficaEstratos",
                p =>
                {
                    p.Add(P("@AnioInicio", filtros.AnioInicio, SqlDbType.Int));
                    p.Add(P("@AnioFin", filtros.AnioFin, SqlDbType.Int));
                    p.Add(P("@CodigoGrupo", filtros.CodigoGrupo, SqlDbType.VarChar, 255));
                    p.Add(P("@CodigoICD", filtros.CodigoICD, SqlDbType.VarChar, 50));
                    p.Add(P("@IdEstado", filtros.IdEstado, SqlDbType.Int));
                    p.Add(P("@IdSexo", filtros.IdSexo, SqlDbType.Int));
                    p.Add(P("@IdRangoEdad", filtros.IdRangoEdad, SqlDbType.Int));
                    p.Add(P("@ClaveInstitucion", filtros.ClaveInstitucion, SqlDbType.NVarChar, 50));
                },
                r => new GraficaEstratoDto
                {
                    EstratoUnidad = SafeStr(r["EstratoUnidad"]),
                    TotalCasos = SafeLong(r["TotalCasos"]),
                    TotalDefunciones = SafeLong(r["TotalDefunciones"])
                });

        // ═══════════════════════════════════════════════════════
        //  GRÁFICA DE SUBGRUPOS  →  sp_Dashboard_GraficaSubgrupos
        //  (no recibe @CodigoICD — compara subcategorías dentro del grupo)
        // ═══════════════════════════════════════════════════════

        public Task<List<GraficaSubgrupoDto>> GetGraficaSubgruposAsync(DashboardFiltrosDto filtros, int topN = 20)
            => SpListAsync("dbo.sp_Dashboard_GraficaSubgrupos",
                p =>
                {
                    p.Add(P("@AnioInicio", filtros.AnioInicio, SqlDbType.Int));
                    p.Add(P("@AnioFin", filtros.AnioFin, SqlDbType.Int));
                    p.Add(P("@CodigoGrupo", filtros.CodigoGrupo, SqlDbType.VarChar, 255));
                    p.Add(P("@IdEstado", filtros.IdEstado, SqlDbType.Int));
                    p.Add(P("@IdSexo", filtros.IdSexo, SqlDbType.Int));
                    p.Add(P("@IdRangoEdad", filtros.IdRangoEdad, SqlDbType.Int));
                    p.Add(P("@ClaveInstitucion", filtros.ClaveInstitucion, SqlDbType.NVarChar, 50));
                    p.Add(P("@Estrato", filtros.Estrato, SqlDbType.NVarChar, 50));
                    p.Add(P("@TopN", topN, SqlDbType.Int));
                },
                r => new GraficaSubgrupoDto
                {
                    CodigoICD = SafeStr(r["codigoICD"]),
                    NombreEnfermedad = SafeStr(r["nombreEnfermedad"]),
                    TotalCasos = SafeLong(r["TotalCasos"]),
                    TotalDefunciones = SafeLong(r["TotalDefunciones"])
                });

        // ═══════════════════════════════════════════════════════
        //  GRÁFICA DÍAS DE ESTANCIA  →  sp_Dashboard_GraficaDiasEstancia
        // ═══════════════════════════════════════════════════════

        public Task<List<GraficaDiasEstanciaDto>> GetGraficaDiasEstanciaAsync(DashboardFiltrosDto filtros)
            => SpListAsync("dbo.sp_Dashboard_GraficaDiasEstancia",
                p => AddFiltrosComunes(p, filtros),
                r => new GraficaDiasEstanciaDto
                {
                    IdRangoEdad = SafeInt(r["idRangoEdad"]),
                    RangoEdad = SafeStr(r["RangoEdad"]),
                    PromedioDiasEstancia = SafeDouble(r["PromedioDiasEstancia"]),
                    TotalCasos = SafeLong(r["TotalCasos"])
                });

        // ═══════════════════════════════════════════════════════
        //  TABLA PAGINADA  →  sp_Dashboard_TablaDatos
        // ═══════════════════════════════════════════════════════

        public Task<TablaDatosResponseDto> GetTablaDatosAsync(TablaDatosRequestDto request)
            => SpTablaPaginadaAsync("dbo.sp_Dashboard_TablaDatos", p =>
            {
                AddFiltrosComunes(p, request);
                p.Add(P("@Pagina", request.Pagina, SqlDbType.Int));
                p.Add(P("@RegistrosPagina", request.RegistrosPagina, SqlDbType.Int));
            });
    }
}

File: plataforma-geointeligente-Front/src/app/core/services/dashboard.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../enviroments/enviroment.development';

// ═══════════════════════════════════════════════════════════
//  Interfaces – catálogos de filtros
// ═══════════════════════════════════════════════════════════

export interface AnioDisponibleDto {
  anio: number;
}

export interface CategoriaEnfermedadDto {
  codigoGrupo: string;
  codigoICD: string;
  nombreEnfermedad: string;
}

export interface SubcategoriaEnfermedadDto {
  codigoICD: string;
  nombreEnfermedad: string;
}

export interface EstadoActivoDto {
  idEstado: number;
  nombreEstado: string;
}

export interface RangoEdadFiltroDto {
  idRangoEdad: number;
  etiquetaRango: string;
}

export interface SexoFiltroDto {
  idSexo: number;
  descripcion: string;
}

export interface InstitucionFiltroDto {
  claveInstitucion: string;
  institucion: string;
}

export interface EstratoFiltroDto {
  estratoUnidad: string;
}

// ═══════════════════════════════════════════════════════════
//  Interfaces – parámetros de filtros compartidos
// ═══════════════════════════════════════════════════════════

export interface DashboardFiltrosParams {
  anioInicio: number;
  anioFin: number;
  codigoGrupo?: string | null;
  codigoICD?: string | null;
  idEstado?: number | null;
  idSexo?: number | null;
  idRangoEdad?: number | null;
  claveInstitucion?: string | null;
  estrato?: string | null;
}

// ═══════════════════════════════════════════════════════════
//  Interfaces – respuestas de datos
// ═══════════════════════════════════════════════════════════

export interface IndicadoresTotalesDto {
  totalCasos: number;
  totalDefunciones: number;
  poblacionBase: number;
  tasaIncidencia: number | null;
  tasaMortalidad: number | null;
}

export interface TendenciaAnualDto {
  anio: number;
  totalCasos: number;
  totalDefunciones: number;
  poblacion: number | null;
  tasaIncidencia: number | null;
  tasaMortalidad: number | null;
}

export interface MapaEstadoDto {
  idEstado: number;
  nombreEstado: string;
  casos: number;
  defunciones: number;
  poblacion: number | null;
  tasaIncidencia: number | null;
  tasaMortalidad: number | null;
}

export interface GraficaEdadDto {
  idRangoEdad: number;
  rangoEdad: string;
  totalCasos: number;
  totalDefunciones: number;
}

export interface GraficaInstitucionDto {
  claveInstitucion: string;
  nombreInstitucion: string;
  totalCasos: number;
  totalDefunciones: number;
}

export interface GraficaEstratoDto {
  estratoUnidad: string;
  totalCasos: number;
  totalDefunciones: number;
}

export interface GraficaSubgrupoDto {
  codigoICD: string;
  nombreEnfermedad: string;
  totalCasos: number;
  totalDefunciones: number;
}

export interface GraficaDiasEstanciaDto {
  idRangoEdad: number;
  rangoEdad: string;
  promedioDiasEstancia: number;
  totalCasos: number;
}

export interface TablaDatosFilaDto {
  idRegistro: number;
  fechaIngreso: string;
  anio: number;
  diasEstancia: number;
  nombreEstado: string;
  nombreMunicipio: string;
  edad: number;
  sexo: string;
  codigoGrupo: string;
  codigoICD: string;
  claveInstitucion: string;
  nombreInstitucion: string;
  estratoUnidad: string;
  idMotivoEgreso: number;
  esDefuncion: number;
}

export interface TablaDatosResponseDto {
  totalRegistros: number;
  datos: TablaDatosFilaDto[];
}

// ═══════════════════════════════════════════════════════════
//  Service
// ═══════════════════════════════════════════════════════════

@Injectable({ providedIn: 'root' })
export class DashboardService {
  private readonly base = `${environment.apiUrl}/api/dashboard`;

  constructor(private http: HttpClient) {}

  // ──────── Catálogos de filtros ────────

  getAniosDisponibles(): Observable<AnioDisponibleDto[]> {
    return this.http.get<AnioDisponibleDto[]>(`${this.base}/filtros/anios-disponibles`);
  }

  getCategoriasEnfermedad(): Observable<CategoriaEnfermedadDto[]> {
    return this.http.get<CategoriaEnfermedadDto[]>(`${this.base}/filtros/categorias-enfermedad`);
  }

  getSubcategoriasEnfermedad(codigoGrupo: string): Observable<SubcategoriaEnfermedadDto[]> {
    const params = new HttpParams().set('codigoGrupo', codigoGrupo);
    return this.http.get<SubcategoriaEnfermedadDto[]>(`${this.base}/filtros/subcategorias-enfermedad`, { params });
  }

  getEstados(): Observable<EstadoActivoDto[]> {
    return this.http.get<EstadoActivoDto[]>(`${this.base}/filtros/estados`);
  }

  getRangosEdad(): Observable<RangoEdadFiltroDto[]> {
    return this.http.get<RangoEdadFiltroDto[]>(`${this.base}/filtros/rangos-edad`);
  }

  getSexos(): Observable<SexoFiltroDto[]> {
    return this.http.get<SexoFiltroDto[]>(`${this.base}/filtros/sexos`);
  }

  getInstituciones(): Observable<InstitucionFiltroDto[]> {
    return this.http.get<InstitucionFiltroDto[]>(`${this.base}/filtros/instituciones`);
  }

  getEstratos(): Observable<EstratoFiltroDto[]> {
    return this.http.get<EstratoFiltroDto[]>(`${this.base}/filtros/estratos`);
  }

  // ──────── Endpoints de datos (con filtros comunes) ────────

  getIndicadores(f: DashboardFiltrosParams): Observable<IndicadoresTotalesDto> {
    return this.http.get<IndicadoresTotalesDto>(`${this.base}/indicadores`, { params: this.toParams(f) });
  }

  getTendencia(f: DashboardFiltrosParams): Observable<TendenciaAnualDto[]> {
    return this.http.get<TendenciaAnualDto[]>(`${this.base}/tendencia`, { params: this.toParams(f) });
  }

  getMapaEstados(f: DashboardFiltrosParams): Observable<MapaEstadoDto[]> {
    return this.http.get<MapaEstadoDto[]>(`${this.base}/mapa-estados`, { params: this.toParams(f) });
  }

  getGraficaEdades(f: DashboardFiltrosParams): Observable<GraficaEdadDto[]> {
    return this.http.get<GraficaEdadDto[]>(`${this.base}/grafica-edades`, { params: this.toParams(f) });
  }

  getGraficaInstituciones(f: DashboardFiltrosParams): Observable<GraficaInstitucionDto[]> {
    return this.http.get<GraficaInstitucionDto[]>(`${this.base}/grafica-instituciones`, { params: this.toParams(f) });
  }

  getGraficaEstratos(f: DashboardFiltrosParams): Observable<GraficaEstratoDto[]> {
    return this.http.get<GraficaEstratoDto[]>(`${this.base}/grafica-estratos`, { params: this.toParams(f) });
  }

  getGraficaSubgrupos(f: DashboardFiltrosParams): Observable<GraficaSubgrupoDto[]> {
    return this.http.get<GraficaSubgrupoDto[]>(`${this.base}/grafica-subgrupos`, { params: this.toParams(f) });
  }

  getGraficaDiasEstancia(f: DashboardFiltrosParams): Observable<GraficaDiasEstanciaDto[]> {
    return this.http.get<GraficaDiasEstanciaDto[]>(`${this.base}/grafica-dias-estancia`, { params: this.toParams(f) });
  }

  getTabla(f: DashboardFiltrosParams, pagina = 1, registrosPagina = 100): Observable<TablaDatosResponseDto> {
    let params = this.toParams(f);
    params = params.set('pagina', pagina.toString()).set('registrosPagina', registrosPagina.toString());
    return this.http.get<TablaDatosResponseDto>(`${this.base}/tabla`, { params });
  }

  // ──────── Helper ────────

  private toParams(f: DashboardFiltrosParams): HttpParams {
    let p = new HttpParams()
      .set('anioInicio', f.anioInicio.toString())
      .set('anioFin', f.anioFin.toString());

    if (f.codigoGrupo) p = p.set('codigoGrupo', f.codigoGrupo);
    if (f.codigoICD) p = p.set('codigoICD', f.codigoICD);
    if (f.idEstado != null) p = p.set('idEstado', f.idEstado.toString());
    if (f.idSexo != null) p = p.set('idSexo', f.idSexo.toString());
    if (f.idRangoEdad != null) p = p.set('idRangoEdad', f.idRangoEdad.toString());
    if (f.claveInstitucion) p = p.set('claveInstitucion', f.claveInstitucion);
    if (f.estrato) p = p.set('estrato', f.estrato);

    return p;
  }
}
